### ISSUE TITLE
Kubernetes e2e suite.[sig-storage] Volume limits should verify that all nodes have volume limits

### ISSUE THREAD

**Issue by olemarkus** [kubernetes-sigs/aws-ebs-csi-driver#915]
/kind bug

**What happened?**

When using the CSI driver, the following tests fail:
```Kubernetes e2e suite.[sig-storage] Volume limits should verify that all nodes have volume limits```

See https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/e2e-kops-grid-scenario-aws-cloud-controller-manager/1400097064578191360

This test is _not_  a part of the CSI migration test suite the EBS CSI driver is using, but is still related to that migration according to this comment: https://github.com/kubernetes/kops/issues/10777#issuecomment-852344969

**What you expected to happen?**

All tests to pass

**How to reproduce it (as minimally and precisely as possible)?**
Running the k/k test suite against a cluster running the CSI driver

**Environment**
- Kubernetes version (use `kubectl version`): v1.22-dev (any version should be affected by this)
- Driver version: v1.0.0

**Comment by wongma7** [comment]
I think it is because the csi driver daemonset pod is not running on the control plane node(s). I haven't verified yet though. The test naively checks all ready&schedulable nodes, that includes control plane nodes.

With migration off, the in-tree EBS driver will set attachable-volumes-aws-ebs. So the kubelet on control plane node will set it.

With migration on, the CSI EBS driver will set attachable-volumes-aws-ebs. So there is nothing on the control plane node to set it.

**Comment by rifelpet** [comment]
[Here](https://storage.googleapis.com/kubernetes-jenkins/logs/e2e-kops-grid-scenario-aws-cloud-controller-manager/1400217873212772352/artifacts/cluster-info/nodes.yaml) is the list of Nodes from a failing test. Note that _none_ of them have `attachable-volumes-aws-ebs` set.


The ebs-csi-node daemonset does run on the control plane nodes. In the yaml above, `ip-172-20-57-200.eu-west-3.compute.internal` is the control plane node and its [kubelet logs](https://storage.googleapis.com/kubernetes-jenkins/logs/e2e-kops-grid-scenario-aws-cloud-controller-manager/1400217873212772352/artifacts/ip-172-20-57-200.eu-west-3.compute.internal/kubelet.log) mention `ebs-csi-node-dvln9` whose logs are [here](https://storage.googleapis.com/kubernetes-jenkins/logs/e2e-kops-grid-scenario-aws-cloud-controller-manager/1400217873212772352/artifacts/cluster-info/kube-system/ebs-csi-node-dvln9/logs.txt). 

None of the [other logs](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/e2e-kops-grid-scenario-aws-cloud-controller-manager/1400217873212772352/artifacts/cluster-info/kube-system/) from ebs-csi-node, ebs-csi-controller, or aws-cloud-controller-manager mention anything about `attachable-volumes-aws-ebs` so I'm struggling to understand why it would not be set on any of the nodes. I suppose we could try explicitly setting [--volume-attach-limit](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/a2a1b1f292ce27d0bd9034e51130d0f1a4b34dc8/cmd/options/node_options.go#L37) on the ebs-csi-node pods to see if that sticks.

**Comment by wongma7** [comment]
My bad, what I said is wrong, the premise that we need to set this annotation when migration is on is wrong: the test failure is misleading because it only checks the annotation, but, we want it to check CSINode.spec.drivers.allocatable

Let me check if the test can be adapted for the migration scenario or if I can write a new one

example CSINode:
```
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  creationTimestamp: "2021-05-27T23:50:52Z"
  name: ip-192-168-49-52.us-west-2.compute.internal
spec:
  drivers:
  - allocatable:
      count: 39
    name: ebs.csi.aws.com
    nodeID: i-0178353fa0a1840a1
    topologyKeys:
    - topology.ebs.csi.aws.com/zone
    - topology.kubern
```

**Comment by wongma7** [comment]
https://github.com/kubernetes/kubernetes/blob/4ad08fa5b611826505fa2892cdce463f12d32b35/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L109

when CSI driver is installed, CSINode gets updated, the annotation key doesn't get set. So I'm not sure why the test expects it to be set. I still don't have a full picture.

**Comment by wongma7** [comment]
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L169

The scheduler knows how to translate in-tree PVs to CSI PVs in case migration is on https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L192. And after it has translated them to CSI, it will use CSINode spec.drivers.allocatable as the limit, not annotation attachable-volumes-aws-ebs

### Conclusion:
- it's expected and acceptable for `Volume limits should verify that all nodes have volume limits to fail when migration is on`, there is no bug in the driver
- there needs to be a migration-equivalent/applicable version of that test to exercise the complicated codepath I just described above (i.e. check that CSINode.spec.drivers.allocatable has been set)

**Comment by olemarkus** [comment]
Thanks for all these details. So there should be a k/k issue on fixing this then?

Meanwhile, we'll filter this particular test on our side.

**Comment by wongma7** [comment]
yes, tracking in  https://github.com/kubernetes/kubernetes/issues/79660#issuecomment-854884112 to write a new test/enhance the existing volume  limits test

**Comment by wongma7** [comment]
https://github.com/kubernetes/kubernetes/pull/103074 has merged, so the 1.22 version of this "should verify that all nodes have volume limits" test will get skipped if Migration is enabled. As a replacement: if storage.testdriver https://github.com/kubernetes/kubernetes/tree/master/test/e2e/storage/external is passed and it supports volume limits then the test called "should verify that all csinodes have volume limits" will run.
